projects/components/src/utils/test/widget-object/angular-widget-object.ts

```diff
@@ -208,6 +208,32 @@ export class TestElement implements Iterable<TestElement> {
         return this.elements[0].injector;
     }
 
+    /**
+     * Private helper function to find the first parent element that matches css selector
+     */
+    private findParents(element: DebugElement, cssSelector: string): DebugElement {
+        if (!element) {
+            return undefined;
```

Comment:
I'd rather return `null`. `undefined` means something that has been created but hasn't been assigned a value. In JS the two basic examples are when you do `let c;` and `a = {}; a.b`

`null` means something that has been assigned an empty value.

This way of searching is really expensive and wasteful because you're querying at each level of the DOM

The easier way to implement what you are doing is using [Element#matches](https://developer.mozilla.org/en-US/docs/Web/API/Element/matches). In pseudo code:

```ts
if (!debugEl) {
   return null;
}
return debugEl.nativeElement.matches(cssSelector) 
    ? debugEl
    : this.findParent(this.debugEl.parent, cssSelector);
```